### ISSUE TITLE
[2.0.0] timeout regression

### ISSUE THREAD

**Issue by doktorstick** [sahlberg/libnfs#223]
Howdy. I've been able to test the latest round of bugfixes and the 2.0.0 release. I've discovered that the `timeout` refactor has regressed the timeout feature. Specifically, it appears that because the timeout is now associated with `pdu`s, that's the only avenue for timing out.

In my specific case, I modified `nfs-cp` to call `nfs_set_timeout(1000); nfs_set_autoreconnect(0)`. I copy a 36 byte local file to the NFS server. Sometimes the server interface is active, other times it is not.

Let me first present the nominal, successful case with my `printf` instrumentation. This is given so you get an idea of where the instrumentation is taking place.
```
rpc_queue_pdu: pdu 0x155a170 timeout 3007951
rpc_timeout_scan: current_time 3007950
rpc_timeout_scan 2: pdu 0x155a170 timeout 3007951
rpc_queue_pdu: pdu 0x155abd0 timeout 3007951
rpc_timeout_scan: current_time 3007950
rpc_timeout_scan 1: pdu 0x155abd0 timeout 3007951
rpc_timeout_scan: current_time 3007950
rpc_timeout_scan 2: pdu 0x155abd0 timeout 3007951
rpc_timeout_scan: current_time 3007950
rpc_queue_pdu: pdu 0x155abd0 timeout 3007951
rpc_timeout_scan: current_time 3007950
rpc_timeout_scan 2: pdu 0x155abd0 timeout 3007951
rpc_queue_pdu: pdu 0x15595b0 timeout 3007951
rpc_timeout_scan: current_time 3007950
rpc_timeout_scan 1: pdu 0x15595b0 timeout 3007951
rpc_timeout_scan: current_time 3007950
rpc_timeout_scan 2: pdu 0x15595b0 timeout 3007951
rpc_timeout_scan: current_time 3007950
rpc_queue_pdu: pdu 0x155a170 timeout 3007951
rpc_timeout_scan: current_time 3007950
rpc_timeout_scan 2: pdu 0x155a170 timeout 3007951
rpc_queue_pdu: pdu 0x15595c0 timeout 3007951
rpc_timeout_scan: current_time 3007950
rpc_timeout_scan 1: pdu 0x15595c0 timeout 3007951
rpc_timeout_scan: current_time 3007950
rpc_timeout_scan 2: pdu 0x15595c0 timeout 3007951
rpc_timeout_scan: current_time 3007950
rpc_queue_pdu: pdu 0x155b1f0 timeout 3007951
rpc_timeout_scan: current_time 3007950
rpc_timeout_scan 2: pdu 0x155b1f0 timeout 3007951
rpc_queue_pdu: pdu 0x1559590 timeout 3007951
rpc_timeout_scan: current_time 3007950
rpc_timeout_scan 1: pdu 0x1559590 timeout 3007951
rpc_timeout_scan: current_time 3007950
rpc_timeout_scan 2: pdu 0x1559590 timeout 3007951
rpc_queue_pdu: pdu 0x15596b0 timeout 3007951
rpc_timeout_scan: current_time 3007950
rpc_timeout_scan 1: pdu 0x15596b0 timeout 3007951
rpc_timeout_scan: current_time 3007950
rpc_timeout_scan 2: pdu 0x15596b0 timeout 3007951
rpc_timeout_scan: current_time 3007950
rpc_queue_pdu: pdu 0x1559560 timeout 3007951
rpc_timeout_scan: current_time 3007950
rpc_timeout_scan 2: pdu 0x1559560 timeout 3007951
rpc_queue_pdu: pdu 0x1559640 timeout 3007951
rpc_timeout_scan: current_time 3007950
rpc_timeout_scan 1: pdu 0x1559640 timeout 3007951
rpc_timeout_scan: current_time 3007950
rpc_timeout_scan 2: pdu 0x1559640 timeout 3007951
rpc_queue_pdu: pdu 0x15598e0 timeout 3007951
rpc_timeout_scan: current_time 3007950
rpc_timeout_scan 1: pdu 0x15598e0 timeout 3007951
rpc_timeout_scan: current_time 3007950
rpc_timeout_scan 2: pdu 0x15598e0 timeout 3007951
rpc_timeout_scan: current_time 3007950
rpc_queue_pdu: pdu 0x1559560 timeout 3007951
rpc_timeout_scan: current_time 3007950
rpc_timeout_scan 2: pdu 0x1559560 timeout 3007951
rpc_timeout_scan: current_time 3007950
copied 36 bytes
rpc_queue_pdu: pdu 0x1559630 timeout 3007951
rpc_timeout_scan: current_time 3007950
rpc_timeout_scan 2: pdu 0x1559630 timeout 3007951
rpc_timeout_scan: current_time 3007950
```

Here's a failure when the interface is in the down state at the start of the test. There are two things to note here: 1) there was no `rpc_queue_pdu` call (so nothing to hang the timeout on); and 2) the timeout was 1000 ms, but it took 3000 ms to detect the failure. 

```
$ sudo /usr/local/bin/nfs-cp ../file.0 nfs://192.168.33.136/exports/share1/file.0
rpc_timeout_scan: current_time 3009037
rpc_timeout_scan: current_time 3009037
[... snipped rpc_timeout_scan lines ...]
rpc_timeout_scan: current_time 3009039
rpc_timeout_scan: current_time 3009039
Failed to mount nfs share : nfs_service failed
Failed to open nfs://192.168.33.136/exports/share1/file.0
```

But it gets worse... This took 127 seconds. The server's interface is cycling up and down every couple of seconds.
```
rpc_queue_pdu: pdu 0x234b170 timeout 3008183
rpc_timeout_scan: current_time 3008182
rpc_timeout_scan 2: pdu 0x234b170 timeout 3008183
rpc_queue_pdu: pdu 0x234bbd0 timeout 3008183
rpc_timeout_scan: current_time 3008182
rpc_timeout_scan 1: pdu 0x234bbd0 timeout 3008183
rpc_timeout_scan: current_time 3008182
rpc_timeout_scan 2: pdu 0x234bbd0 timeout 3008183
rpc_timeout_scan: current_time 3008182
rpc_timeout_scan: current_time 3008182
[... snipped rpc_timeout_scan lines ...]
rpc_timeout_scan: current_time 3008309
rpc_timeout_scan: current_time 3008309
Failed to mount nfs share : nfs_service failed
Failed to open nfs://192.168.33.136/exports/share1/file.0...
```

I thought I would be cheeky and add a flag to the `rpc` structure (`has_timed_out`) and in `wait_for_reply` set that if `poll (nfs_get_timeout())` returned 0 and allow `rpc_timeout_scan` at the very bottom do:
```
if (rpc->has_timed_out) {
    rpc_set_error (rpc, "command timed out");
}
```
but while it timed out almost exactly at the right moment, it eventually `SIGSEGV` (likely due to the genesis bug of the timeout rework).

**Comment by doktorstick** [comment]
Here's two operations where it hangs:
```
nfs_mount
nfs_mount_async
rpc_connect_program_async
rpc_connect_async
rpc_connect_sockaddr_async
rpc_connect_sockaddr_async do loop...
rpc_connect_sockaddr_async binding...
rpc_connect_sockaddr_async binding... rc=0
rpc_connect_sockaddr_async ret 0
nfs_get_rpc_context
rpc_timeout_scan: current_time 3016401
rpc_timeout_scan: current_time 3016403
...
```
```
...
rpc_connect_program_3_cb
rpc_disconnect
rpc_connect_async
rpc_connect_sockaddr_async
rpc_connect_sockaddr_async do loop...
rpc_connect_sockaddr_async binding...
rpc_connect_sockaddr_async binding... rc=0
rpc_connect_sockaddr_async ret 0
rpc_timeout_scan: current_time 3016538
rpc_timeout_scan: current_time 3016540
...
```

**Comment by doktorstick** [comment]
The problem may be that `rpc_connect_sockaddr_async` may take a while for the socket operations to succeed/fail, and so an event on the `fd` isn't generated. Since there's no external mechanism aborting prolonged transactions, it hangs. Here's more detailed instrumentation (the dots ('.') represent the `wait_for_nfs_reply` loop @ 2sec/dot):

```
nfs_mount nfs_get_rpc_context (nfs=0x704070)
nfs_mount nfs_mount_async (nfs=0x704070, server=0x7083e0 "192.168.33.136", export=0x708410 "/exports/share1", cb=0x7ffff7ba74c8 <mount_cb>, private_data=0x7fffffffe9e0)
nfs_mount_async rpc_connect_program_async (rpc=0x7040e0, server=0x7083e0 "192.168.33.136", program=100005, version=3, cb=0x7ffff7b9ab36 <nfs_mount_5_cb>, private_data=0x708450)
rpc_connect_program_async rpc_connect_async (rpc=0x7040e0, server=0x7083e0 "192.168.33.136", port=111, cb=0x7ffff7b995b9 <rpc_connect_program_1_cb>, private_data=0x708560)
rpc_connect_async rpc_set_sockaddr (rpc=0x7040e0, server=0x7083e0 "192.168.33.136", port=111)
rpc_connect_async rpc_connect_sockaddr_async (rpc=0x7040e0)
rpc_connect_sockaddr_async create_socket (domain=2, type=1, protocol=6)
rpc_connect_sockaddr_async set_bind_device (fd=4, ifname=0x708308 "")
rpc_connect_sockaddr_async rpc_current_time ()
rpc_connect_sockaddr_async set_nonblocking (fd=4)
rpc_connect_sockaddr_async set_nolinger (fd=4)
nfs_mount wait_for_nfs_reply (nfs=0x704070, cb_data=0x7fffffffe9e0)......................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................
wait_for_nfs_reply nfs_get_fd (nfs=0x704070)
nfs_get_fd rpc_get_fd (rpc=0x7040e0)
wait_for_nfs_reply nfs_which_events (nfs=0x704070)
nfs_which_events rpc_which_events (rpc=0x7040e0)
rpc_which_events rpc_has_queue (q=0x704120)
wait_for_nfs_reply nfs_service (nfs=0x704070, revents=28)
nfs_service rpc_service (rpc=0x7040e0, revents=28)
rpc_service rpc_set_error (rpc=0x7040e0, error_string=0x7ffff7bc6b48 "rpc_service: socket error %s(%d).")
rpc_service rpc_set_error (rpc=0x7040e0, error_string=0x7ffff7bc6b9c "Socket failed with POLLHUP")
rpc_service maybe_call_connect_cb (rpc=0x7040e0, status=1)
maybe_call_connect_cb rpc_connect_program_1_cb (rpc=0x7040e0, status=1, command_data=0x709ba0, private_data=0x708560)
rpc_connect_program_1_cb nfs_mount_5_cb (rpc=0x7040e0, status=1, command_data=0x709ba0, private_data=0x708450)
nfs_mount_5_cb check_nfs_error (nfs=0x704070, status=1, data=0x708450, command_data=0x709ba0)
check_nfs_error mount_cb (status=-14, nfs=0x704070, data=0x709ba0, private_data=0x7fffffffe9e0)
mount_cb nfs_set_error (nfs=0x704070, error_string=0x7ffff7bc5900 "mount/mnt call failed with \"%s\"")
nfs_mount_5_cb free_nfs_cb_data (data=0x708450)
rpc_connect_program_1_cb free_rpc_cb_data (data=0x708560)
wait_for_nfs_reply nfs_set_error (nfs=0x704070, error_string=0x7ffff7bc58ed "nfs_service failed")
nfs_mount rpc_disconnect (rpc=0x7040e0, error=0x7ffff7bc5937 "failed mount")
open_file nfs_get_error (nfs=0x704070)
nfs_get_error rpc_get_error (rpc=0x7040e0)
open_file free_file_context (file_context=0x704040)
free_file_context nfs_destroy_context (nfs=0x704070)
nfs_destroy_context rpc_destroy_context (rpc=0x7040e0)
rpc_destroy_context rpc_purge_all_pdus (rpc=0x7040e0, status=2, error=0x0)
```

**Comment by sahlberg** [comment]
Try latest master,  it will handle PDU timeouts a little better when the link/socket is failing.

Previously, suding socket errors, it might abort rpc_service() before it got to the stage where it would scan for and abort any pdus that had timed out, casing potentially a long delay before the pdus were aborted.

It will now do this check at the beginning of rpc_service() and thus will timeout the PDUs quicker, even if the socket is in errored state for a prolonged time.

The socket reconnect itself is a different story. It may take as long as it takes which could be unbounded in the case of a network outage.


For your tests, a possibly better test than bringing the interface up/down (which affects other things as well such as the dns resolver...) may be to leave the interface up and instead just use
iptables to block/allow traffic to <server>:2049:TCP

**Comment by doktorstick** [comment]
_The socket reconnect itself is a different story. It may take as long as it takes which could be unbounded in the case of a network outage. :-(_

Hrm, that's not good. That's what was great about the old timeout mechanism (bug not withstanding). Right now, with the current master, timeout seems to be working for everything except connect (`nfs_service failed`). I'll have to tackle the connect problem if `libnfs` timeout behavior is at an acceptable level for you; I can't have it hanging indefinitely.

As for my tests, to be clear it's the server interface (no DNS entry) I'm taking down (simulating network outages or partitions that reconverge). `iptables` might be a swell way to do it, provided it black-holes the packet and the client can't distinguish between the two cases. (I think it can do that; my `iptables` knowledge is ephemeral and I've slept since I last used it.)

**Comment by doktorstick** [comment]
Double hrm. I need amend that previous post. With a stable server connection, I was running looping tests for copying a 36 byte file and it would occasionally timeout (~200 copies). Looking through the code, I see that you are using only the seconds portion of `clock_gettime`. This isn't good enough resolution. If it samples at t=0.999999, the `pdu->timeout = 0 + 1000/1000 = 1`, only 0.000001 second away.

I changed it to to use millisecond resolution and it is rock solid.
```
--- a/include/libnfs-private.h
+++ b/include/libnfs-private.h
@@ -175,7 +175,7 @@ struct rpc_pdu {
 #define PDU_DISCARD_AFTER_SENDING 0x00000001
         uint32_t flags;

-       time_t timeout;
+       long timeout;
 };

 void rpc_reset_queue(struct rpc_queue *q);
@@ -233,7 +233,7 @@ int rpc_get_timeout(struct rpc_context *rpc);
 int rpc_add_fragment(struct rpc_context *rpc, char *data, uint32_t size);
 void rpc_free_all_fragments(struct rpc_context *rpc);
 int rpc_is_udp_socket(struct rpc_context *rpc);
-int rpc_current_time(void);
+long rpc_current_time(void);

 void *zdr_malloc(ZDR *zdrs, uint32_t size);

diff --git a/lib/init.c b/lib/init.c
index 2bd47b1..3a5c533 100644
--- a/lib/init.c
+++ b/lib/init.c
@@ -50,15 +50,15 @@
 #include "libnfs-raw.h"
 #include "libnfs-private.h"

-int rpc_current_time(void)
+long rpc_current_time(void)
 {
 #ifdef HAVE_CLOCK_GETTIME
        struct timespec tp;

        clock_gettime(CLOCK_MONOTONIC_COARSE, &tp);
-       return tp.tv_sec;
+       return tp.tv_sec * 1000 + tp.tv_nsec / 1000000;
 #else
-       return time(NULL);
+       return (long) time(NULL) * 1000;
 #endif
 }

diff --git a/lib/pdu.c b/lib/pdu.c
index 8186f2e..5b4191b 100644
--- a/lib/pdu.c
+++ b/lib/pdu.c
@@ -223,7 +223,7 @@ int rpc_queue_pdu(struct rpc_context *rpc, struct rpc_pdu *pdu)
        assert(rpc->magic == RPC_CONTEXT_MAGIC);

        if (rpc->timeout > 0) {
-               pdu->timeout = rpc_current_time() + rpc->timeout / 1000;
+               pdu->timeout = rpc_current_time() + rpc->timeout;
        } else {
                pdu->timeout = 0;
        }
```

**Comment by sahlberg** [comment]
Awesome patch!  Do you want to send it as a pull request so you get the proper credits for you ?
If not, I can just apply it by hand.


Just bear with me. We WILL get the timeout handling into a state where it will work for your use case.
I will start trying to look at the connect issue.

**Comment by tokibito** [comment]
Is not this patch yet integrated?
We are waiting for this problem to be solved.

**Comment by sahlberg** [comment]
Please send as a pull request so I can review.

**Comment by sahlberg** [comment]
Do we still need the patch in : https://github.com/sahlberg/libnfs/issues/223#issuecomment-312728789

master already contain a change that should be equivalent to that patch.

**Comment by doktorstick** [comment]
No. You merged my pull request on 2017-07-07 (#224).
